framework/auth/views.py

```diff
@@ -257,6 +301,64 @@ def confirm_email_get(token, auth=None, **kwargs):
     ))
 
 
+@collect_auth
+@must_be_logged_in
+def unconfirmed_email_remove(auth=None):
+    """Called at login if user cancels their merge or email add.
+    methods: DELETE
+    """
+    user = auth.user
+    json_body = request.get_json()
+    try:
+        given_token = json_body['token']
+    except KeyError:
+        raise HTTPError(http.BAD_REQUEST, data={
+            'message_short': 'Invalid token',
+            'message_long': 'The token provided is invalid'
+        })
+    user.clean_email_verifications(given_token=given_token)
+    user.save()
+    return {
+        'status': 'success',
+        'removed_email': json_body['address']
+    }, 200
+
+
+@collect_auth
+@must_be_logged_in
+def unconfirmed_email_add(auth=None):
+    """Called at login if user confirms their merge or email add.
+    methods: PUT
+    """
+    user = auth.user
+    json_body = request.get_json()
+    try:
+        token = json_body['token']
+    except KeyError:
+        raise HTTPError(http.BAD_REQUEST, data={
+            'message_short': 'Invalid token',
+            'message_long': 'The token provided is invalid'
+        })
+    try:
+        user.confirm_email(token, merge=True)
+    except exceptions.EmailConfirmTokenError as e:
+        raise HTTPError(http.BAD_REQUEST, data={
+            'message_short': e.message_short,
+            'message_long': e.message_long
+        })
+
+    if token:
```

Comment:
Won't this always be `True`? You're ensuring that token is provided ~L336. If this is the case you should be able to remove this conditional.